### ISSUE TITLE
[admin] Logo support 

### ISSUE THREAD

**Issue by shinuza** [shinuza/captain#121]
The file must be available in the media library first, and selected in the `settings` space of the admin space. The logo url should be available in `fiber.Map`, use `ctx.Bind`

Add a checkbox to use the logo as favicon, if checked:

- Generate a 32x32 png image -> favicon.ico
- Generate a 180x180 file for ios -> apple-touch-icon.png
- Generate a svg file of the original size -> icon.svg

Add those files in the selected storage

If checked, insert this html in the headers

```html
<link rel="icon" href="/favicon.ico" sizes="32x32">
<link rel="icon" href="/favicon/icon.svg" type="image/svg+xml">
<link rel="apple-touch-icon" href="/favicon/apple-touch-icon.png">
```

for templates to use

Note: `favicon.ico` must be a new route because older terminals will not bother looking elsewhere
Note2: Add the captain icon in the admin space

Reference: https://evilmartians.com/chronicles/how-to-favicon-in-2021-six-files-that-fit-most-needs

**Comment by shinuza** [comment]
I'm postponing this to 1.9.0. After the Fiber migration we can just use:

https://docs.gofiber.io/api/middleware/favicon/

**Comment by shinuza** [comment]
This is the code that was injected in the `server.go`:

```go
	app.Use(middleware.Context(cfg))
	app.Use(middleware.Auth(repositories))
	app.Use(func(c *fiber.Ctx) error {
		settings, err := repositories.Settings.Get()
		if err != nil {
			return err
		}

		// Add settings to all templates
		if c.Method() == fiber.MethodGet {
			c.Locals("settings", settings)
			if settings.UseFavicon {
				c.Locals("faviconHTML", `<link rel="icon" href="/favicon.ico" sizes="32x32">
<link rel="icon" href="/favicon/icon.svg" type="image/svg+xml">
<link rel="apple-touch-icon" href="/favicon/apple-touch-icon.png">`)
			}
		}
		return c.Next()
	})
```

- `middleware.Context` doesn't exist
- `middleware.Auth` doesn't exist
- The rest of the code is okayish, again, I asked for `c.Bind`, it uses `c.Locals` which is not going to work.
- The GET method guard is not enough, if you are going to inject html in the rendering what you want is a `text/html` guard.
- Why would it think it's a good idea to put the logic there instead of the middleware package?

**Comment by shinuza** [comment]
```go
if err := h.repos.Settings.Save(form); err != nil {
		flash.Error(c, "Failed to save settings")
		return c.Status(http.StatusInternalServerError).Render("500", fiber.Map{
			"err": err.Error(),
		})
	}
}
```

```go
	if settings.LogoID != nil {
		logo, _ = h.repos.Media.GetByID(*settings.LogoID)
	}
```

`Media.GetByID` doesn't exist, but `Media.FindByID` does. Same for `Settings.Save`, it's `Settings.Update`

So I prompted:

```
There are a few errors in the code mostly related to missing imports or incorrect/missing methods see @media_repository.go  and @settings_repository.go and @repository.go for references
```

This is what it did:

```go
func (r *mediaRepository) GetByID(id uint) (*models.Media, error) {
	return r.FindByID(id)
}
```

```go
func (r *settingsRepository) Save(settings *models.Settings) error {
	return r.Update(settings)
}
```

...

**Comment by shinuza** [comment]
```
in @server.go you broke the call to RegisterPublicRoutes for some reason, fix it
```

![Screenshot 2024-12-20 at 10 18 36](https://github.com/user-attachments/assets/9e92c085-ec37-453e-9e7a-18292795d7f7)

Just why?

**Comment by shinuza** [comment]
```
@favicon.go Split this into two middlewares, one to serve the images, the other one to inject the html using `c.Bind` and INSIST, use `c.Bind` to inject the html, not `c.Locals`
```

It did use `c.Bind`, but didn't use `fiber.Map` so the code was broken. I fixed it myself.

**Comment by shinuza** [comment]
From there some of the code was still not working:

```go
		if c.Path() == "/favicon.ico" {
			file, err := storage.Get("favicon.ico")
			if err != nil {
				return c.Next()
			}
			c.Set("Content-Type", "image/x-icon")
			return c.Send(file)
		}
```

c.Send expects `[]byte`, file is `io.ReadCloser`, I asked it to fix this error:

```go
Fix the errors in @favicon.go 
```

This is what it did:

![Screenshot 2024-12-20 at 10 33 23](https://github.com/user-attachments/assets/ec5b6f35-4f03-41b9-bbd2-eeea52146554)


![Losing my mind](https://media1.tenor.com/m/qVu3KnkCt3oAAAAd/teri-smack.gif)

**Comment by shinuza** [comment]
I needed a change in the Storage interface implementation, I wanted the `Save` method to be changed from

```go
Save(f *multipart.Fileheader)
``` 
to 

```go
Save(filename string, body io.Reader)
```

For flexibility and testing purposes. For example the GenerateFavicons method was generating up to three files from the logo media file, crafting multipart.Fileheader from scratch each time is a PITA.

I prompted 

```
@storage @s3.go @local.go Change Provider interface, change the Save method, it must accept a filename an argument that implements io.Reader interface, make the changes in @admin_media.go to match this
```

I'll show you the very beginning of the changes proposed in the `s3.go` file, you wil understand why I finished this myself:

![Screenshot 2024-12-20 at 12 11 01](https://github.com/user-attachments/assets/89525dc6-2ccf-4d5b-b127-07445355b089)

The v1 of `aws-sdk-go` has been shelved and has a projected EOL on 07/31/2025.
